numba/parfor.py

```diff
@@ -268,6 +269,85 @@ def _analyze_parfor(parfor, equiv_set, typemap, array_analysis):
 
 array_analysis.array_analysis_extensions[Parfor] = _analyze_parfor
 
+
+class PreParforPass(object):
+    """Preprocessing for the Parfor pass. It mostly inlines parallel
+    implementations of numpy functions if available.
+    """
+    def __init__(self, func_ir, typemap, calltypes, typingctx, options):
+        self.func_ir = func_ir
+        self.typemap = typemap
+        self.calltypes = calltypes
+        self.typingctx = typingctx
+        self.options = options
+
+    def run(self):
+        """Run pre-parfor processing pass.
+        """
+        # e.g. convert A.sum() to np.sum(A) for easier match and optimization
+        canonicalize_array_math(self.func_ir, self.typemap,
+                                self.calltypes, self.typingctx)
+        if self.options.numpy:
+            self._replace_parallel_functions(self.func_ir.blocks)
+        self.func_ir.blocks = simplify_CFG(self.func_ir.blocks)
+
+    def _replace_parallel_functions(self, blocks):
+        """
+        Replace functions with their parallel implemntation in
+        replace_functions_map if available.
+        The implementation code is inlined to enable more optimization.
+        """
+        from numba.inline_closurecall import inline_closure_call
+        work_list = list(blocks.items())
+        while work_list:
+            label, block = work_list.pop()
+            for i, instr in enumerate(block.body):
+                if isinstance(instr, ir.Assign):
+                    lhs = instr.target
+                    expr = instr.value
+                    if isinstance(expr, ir.Expr) and expr.op == 'call':
+                        def replace_func():
+                            func_def = get_definition(self.func_ir, expr.func)
+                            callname = find_callname(self.func_ir, expr)
+                            repl_func = replace_functions_map.get(callname, None)
+                            require(repl_func != None)
+                            typs = tuple(self.typemap[x.name] for x in expr.args)
+                            new_func =  repl_func(*typs)
+                            require(new_func != None)
+                            g = copy.copy(self.func_ir.func_id.func.__globals__)
+                            g['numba'] = numba
+                            g['np'] = numpy
+                            # inline the parallel implementation
+                            inline_closure_call(self.func_ir, g,
+                                            block, i, new_func, self.typingctx, typs,
+                                            self.typemap, self.calltypes, work_list)
+                            return True
+                        if guard(replace_func):
+                            break
+                    elif isinstance(expr, ir.Expr) and expr.op == 'getattr':
```

Comment:
What is the purpose of this array.dtype simplification? Any use cases?

Reply:
With things like `a = np.empty(10, b.dtype)` where `b` is produced from a previous parfor, it is useful to have `b.dtype` replaced by the actual type because it is statically known. It used to be that this `b.dtype` statement would get in between two parfors, but since you have introduced `init_prange()` this has become less of an issue, although it would still benefit the second parfor by not having a depdency to variable `b`.

Reply:
Makes sense, thanks.